canary/scripts/test_hmr.py

```diff
@@ -0,0 +1,92 @@
+from openforcefield.topology import Molecule
+from openmmforcefields.generators import SystemGenerator
+from pkg_resources import resource_filename
+from simtk import openmm, unit
+from simtk.openmm import app
+
+DATA_PATH = Path(resource_filename("openforcefields", "../canary/data/")).resolve()
+
+coverage_mols = DATA_PATH / "coverage.smi"
+propyne_mols = DATA_PATH / "propynes.smi"
+
+
+class NANEnergyError(Exception):
+    """Base exception for a system with NaN-like potential energy"""
+
+
+class CanaryError(Exception):
+    """Base exception for canary"""
+
+
+class HMRCanaryError(CanaryError):
+    """Exception for an HMR canary test failing"""
+
+
+def hmr_driver(mol, ff_name):
+    """Given an OpenFF Molecule, run a short 4 fs HMR simulation. This function is adapted from
+    https://github.com/openforcefield/openforcefields/issues/19#issuecomment-689816995"""
+    print(
+        f"Running HMR with force field {ff_name} and molecule with SMILES {mol.to_smiles()}"
+    )
+
+    forcefield_kwargs = {
+        "constraints": app.HBonds,
+        "rigidWater": True,
+        "removeCMMotion": False,
+        "hydrogenMass": 4
+        * unit.amu,  # Does this also _subtract_ mass from heavy atoms?:w
+    }
+
+    system_generator = SystemGenerator(
+        small_molecule_forcefield=ff_name,
+        forcefield_kwargs=forcefield_kwargs,
+        molecules=mol,
+    )
+    system = system_generator.create_system(mol.to_topology().to_openmm())
+
+    temperature = 300 * unit.kelvin
+    collision_rate = 1.0 / unit.picoseconds
+    timestep = 4.0 * unit.femtoseconds
+
+    integrator = openmm.LangevinIntegrator(temperature, collision_rate, timestep)
+    context = openmm.Context(system, integrator)
+    mol.generate_conformers(n_conformers=1)
+    context.setPositions(mol.conformers[0])
+
+    # Run for 10 ps
+    integrator.step(500)
+
+    state = context.getState(getEnergy=True)
+    pot = state.getPotentialEnergy()
+    # OpenMM will silenty "fail" if energies aren't explicitly checked
+    if np.isnan(pot / pot.unit):
+        raise NANEnergyError()
+
+
+if __name__ == "__main__":
+    """This function expects to be called with a list of OFFXML files passed to it,
+    i.e. piped from git diff upstream/master  --name-only"""
+    # Read force field filenames from stdin
+    for line in sys.stdin:
+        if "_unconstrained" in line:
+            continue
+        ff_name = line.split("/")[-1][:-8]
+        # Molecule.from_file fails on pathlib.Path ojects, despite being str-like
+        hmr_mols = Molecule.from_file(
+            str(propyne_mols), file_format="smi", allow_undefined_stereo=True
+        )
+        # TODO: Add coverage set, with known failures stripped out
+        # hmr_mols = Molecule.from_file(str(coverage_mols), file_format='smi', allow_undefined_stereo=True)
+        failed_runs = []
+        for mol in hmr_mols:
+            try:
+                hmr_driver(mol, ff_name)
+            except NANEnergyError:
```

Comment:
On linux, `NaN` failures don't seem to be getting caught. Not a huge deal since failing is the whole point of the canary, but it may be helpful to make it so that multiple NaNs can be caught and a summary of failures can be shown to users.

Reply:
Unfortunately, we run into one of those "un-Python" behaviors in OpenMM, as the triggered exception cannot be caught.

```python3
>>> from simtk.openmm import OpenMMException
>>> try:
...     raise OpenMMException
... except OpenMMException:
...     print('caught!')
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: exceptions must derive from BaseException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
TypeError: catching classes that do not inherit from BaseException is not allowed
```

 I can just catch it later, but not with the specificity we'd normally prefer (catching more than less is better for these tests, anyway):
```
(test) root@af8cfd33010c:/openforcefields# echo openforcefields/offxml/openff-1.2.0.offxml | python canary/scripts/test_hmr.py -vs
Warning: Unable to load toolkit 'OpenEye Toolkit'. The Open Force Field Toolkit does not require the OpenEye Toolkits, and can use RDKit/AmberTools instead. However, if you have a valid license for the OpenEye Toolkits, consider installing them for faster performance and additional file format support: https://docs.eyesopen.com/toolkits/python/quickstart-python/linuxosx.html OpenEye offers free Toolkit licenses for academics: https://www.eyesopen.com/academic-licensing
[16:05:40] WARNING: no name column found on line 0
[16:05:40] WARNING: no name column found on line 1
Running HMR with force field openff-1.2.0 and molecule with SMILES [H][C]#[C][C]([H])([H])[H]
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
Running HMR with force field openff-1.2.0 and molecule with SMILES [H][c]1[c]([H])[c]([H])[c]([C]#[C][C]([H])([H])[H])[c]([H])[c]1[H]
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
/bin/bash: /opt/conda/envs/test/lib/libtinfo.so.6: no version information available (required by /bin/bash)
Traceback (most recent call last):
  File "canary/scripts/test_hmr.py", line 94, in <module>
    raise HMRCanaryError(failed_runs)
__main__.HMRCanaryError: [[Molecule with name '' and SMILES '[H][C]#[C][C]([H])([H])[H]', 'openff-1.2.0', 'NaN Position(s)'], [Molecule with name '' and SMILES '[H][c]1[c]([H])[c]([H])[c]([C]#[C][C]([H])([H])[H])[c]([H])[c]1[H]', 'openff-1.2.0', 'NaN Position(s)']]
```